paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/SchemaAwareStoreWriteOperator.java

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.sink.cdc;
+
+import org.apache.paimon.cdc.Record;
+import org.apache.paimon.data.GenericRow;
+import org.apache.paimon.flink.sink.AbstractStoreWriteOperator;
+import org.apache.paimon.flink.sink.LogSinkFunction;
+import org.apache.paimon.flink.sink.StoreSinkWrite;
+import org.apache.paimon.options.ConfigOption;
+import org.apache.paimon.options.ConfigOptions;
+import org.apache.paimon.schema.TableSchema;
+import org.apache.paimon.table.FileStoreTable;
+import org.apache.paimon.table.sink.SinkRecord;
+import org.apache.paimon.types.DataType;
+import org.apache.paimon.utils.TypeUtils;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Map;
+
+/**
+ * An {@link AbstractStoreWriteOperator} which is aware of schema changes.
+ *
+ * <p>When the input {@link Record} contains a field name not in the current {@link TableSchema}, it
+ * periodically queries the latest schema, until the latest schema contains that field name.
+ */
+public class SchemaAwareStoreWriteOperator extends AbstractStoreWriteOperator<Record> {
+
+    static final ConfigOption<Duration> SCHEMA_UPDATE_WAIT_TIME =
+            ConfigOptions.key("cdc.schema-update-wait-time")
+                    .durationType()
+                    .defaultValue(Duration.ofMillis(500));
+
+    private final long schemaUpdateWaitMillis;
+
+    public SchemaAwareStoreWriteOperator(
+            FileStoreTable table,
+            @Nullable LogSinkFunction logSinkFunction,
+            StoreSinkWrite.Provider storeSinkWriteProvider) {
+        super(table, logSinkFunction, storeSinkWriteProvider);
+        schemaUpdateWaitMillis =
+                table.options().toConfiguration().get(SCHEMA_UPDATE_WAIT_TIME).toMillis();
+    }
+
+    @Override
+    protected SinkRecord processRecord(Record record) throws Exception {
+        Map<String, String> fields = record.fields();
+
+        if (!schemaMatched(fields)) {
+            while (true) {
+                table = table.copyWithLatestSchema();
+                if (schemaMatched(fields)) {
+                    break;
+                }
+                Thread.sleep(schemaUpdateWaitMillis);
+            }
+            write.replace(commitUser -> table.newWrite(commitUser));
+        }
+
+        TableSchema schema = table.schema();
+        GenericRow row = new GenericRow(schema.fields().size());
+        row.setRowKind(record.kind());
+        for (Map.Entry<String, String> field : fields.entrySet()) {
+            String key = field.getKey();
+            String value = field.getValue();
+            int idx = schema.fieldNames().indexOf(key);
+            DataType type = schema.fields().get(idx).type();
+            row.setField(idx, TypeUtils.castFromString(value, type));
```

Comment:
`TypeUtils.castFromString` is limited, can not support complex types.
Add a TODO here?